pkg/vm/engine/tae/txn/txnbase/logtail.go

```diff
@@ -0,0 +1,222 @@
+	size := atomic.LoadInt32(l.activeSize)
+	l.activePage[size] = txn
+	newsize := atomic.AddInt32(l.activeSize, 1)
+
+	if newsize == l.pageSize {
+		l.Lock()
+		defer l.Unlock()
+		txns := l.activePage
+		minTs := txns[0].GetPrepareTS()
+		if minTs.IsEmpty() {
+			minTs = l.minTs
+		}
+		newPage := &txnPage{
+			minTs: minTs,
+			txns:  txns,
+		}
+		l.pages.Set(newPage)
+
+		l.minTs = l.tsAlloc.Alloc()
+		l.activePage = make([]txnif.AsyncTxn, l.pageSize)
+		atomic.StoreInt32(l.activeSize, 0)
+	}
+}
+
+// GetLogtailView returns a read only view of txns at call time.
+// this is cheap operation, the returned view can be accessed without any locks
+func (l *LogtailMgr) GetLogtailView(start, end types.TS, tid uint64) *LogtailView {
+	l.Lock()
+	size := atomic.LoadInt32(l.activeSize)
+	activeView := l.activePage[:size]
+	btreeView := l.pages.Copy()
+	l.Unlock()
+	return &LogtailView{
+		start:      start,
+		end:        end,
+		tid:        tid,
+		btreeView:  btreeView,
+		activeView: activeView,
+	}
+}
+
+// a read only view of txns
+type LogtailView struct {
+	start, end types.TS                 // included
+	tid        uint64                   // table id
+	btreeView  *btree.Generic[*txnPage] // read only btree
+	activeView []txnif.AsyncTxn         // read only active page
+}
+
+func (v *LogtailView) GetDirtyPoints() Dirties {
+	// if tree[segID] is nil, it means that is just a seg operation
+	tree := make(map[uint64]map[uint64]struct{}, 0)
+	var blkSet map[uint64]struct{}
+	var exist bool
+	f := func(txn txnif.AsyncTxn) (moveOn bool) {
+		store := txn.GetStore()
+		if store.HasTableDataChanges(v.tid) {
+			pointSet := store.GetTableDirtyPoints(v.tid)
+			for dirty := range pointSet {
+				if dirty.BlkID == 0 {
+					// a segment operation
+					if _, exist = tree[dirty.SegID]; !exist {
+						tree[dirty.SegID] = nil
+					}
+				} else {
+					// merge the dirty block
+					blkSet, exist = tree[dirty.SegID]
+					if !exist || blkSet == nil {
+						blkSet = make(map[uint64]struct{})
+						tree[dirty.SegID] = blkSet
+					}
+					blkSet[dirty.BlkID] = struct{}{}
+				}
+			}
+		}
+		return true
+	}
+	v.scanTxnBetween(v.start, v.end, f)
+
+	segs := make([]dirtySeg, 0, len(tree))
+	for sig, blkSet := range tree {
+		blk := make([]uint64, 0, len(blkSet))
+		for b := range blkSet {
+			blk = append(blk, b)
+		}
+		segs = append(segs, dirtySeg{Sig: sig, Blks: blk})
+	}
+	return Dirties{Segs: segs}
+}
+
+// Dirties has determined iter order
+type Dirties struct {
+	Segs []dirtySeg
+}
+
+type dirtySeg struct {
+	Sig  uint64
+	Blks []uint64
+}
+
+func (v *LogtailView) HasCatalogChanges() bool {
+	changed := false
+	f := func(txn txnif.AsyncTxn) (moveOn bool) {
+		if txn.GetStore().HasCatalogChanges() {
+			changed = true
+			return false
+		}
+		return true
+	}
+	v.scanTxnBetween(v.start, v.end, f)
+	return changed
+}
+
+func (v *LogtailView) scanTxnBetween(start, end types.TS, f func(txn txnif.AsyncTxn) (moveOn bool)) {
+	var pivot types.TS
```

Comment:
I don't think separating the active page from the others is a good idea. Put them together into the btree.

Reply:
it will be moved to btree to simplify iteration later, for now it is just an inner interface so let it be. I wrote a TODO in comment